owrrm/src/main/java/com/facebook/openwifi/rrm/rca/modules/StationPinger.java

```diff
@@ -0,0 +1,267 @@
+import com.facebook.openwifi.rrm.rca.RCAUtils;
+import com.facebook.openwifi.rrm.rca.RCAUtils.PingResult;
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * Ping service to measure latency/jitter between Wi-Fi APs and clients
+ * (stations).
+ * <p>
+ * This class subscribes to the Kafka "state" topic to retrieve the list of APs
+ * with connected clients, then issues ping commands for each (AP, STA) pair at
+ * a given frequency. All actions are submitted to an executor during Kafka
+ * callbacks.
+ */
+public class StationPinger {
+	private static final Logger logger =
+		LoggerFactory.getLogger(StationPinger.class);
+
+	/** Drop records older than this interval (in ms). */
+	private static final long STATE_STALE_THRESHOLD_MS = 300000; // 5 min
+
+	/** The module parameters. */
+	private final StationPingerParams params;
+
+	/** The uCentral client. */
+	private final UCentralClient client;
+
+	/** The executor service instance. */
+	private final ExecutorService executor;
+
+	/** The Gson instance. */
+	private final Gson gson = new Gson();
+
+	/**
+	 * Map from device (serial number) to the latest map of STAs
+	 * (i.e. client MAC address to Client structure).
+	 */
+	private Map<String, Map<String, State.Interface.Client>> deviceToClients =
+		new ConcurrentHashMap<>();
+
+	/**
+	 * Map of last ping timestamps, keyed on
+	 * {@link #getDeviceKey(String, String)}.
+	 */
+	private Map<String, Long> lastPingTsMap = new ConcurrentHashMap<>();
+
+	/** Constructor. */
+	public StationPinger(
+		StationPingerParams params,
+		UCentralClient client,
+		UCentralKafkaConsumer consumer
+	) {
+		this.params = params;
+		this.client = client;
+		this.executor =
+			Executors.newFixedThreadPool(
+				params.executorThreadCount,
+				new Utils.NamedThreadFactory(
+					"RCA_" + this.getClass().getSimpleName()
+				)
+			);
+
+		if (params.pingIntervalSec < 1) {
+			logger.info("StationPinger is disabled");
+			return; // quit before registering listeners
+		}
+
+		// Register Kafka listener
+		if (consumer != null) {
+			consumer.addKafkaListener(
+				getClass().getSimpleName(),
+				new UCentralKafkaConsumer.KafkaListener() {
+					@Override
+					public void handleStateRecords(List<KafkaRecord> records) {
+						handleKafkaStateRecords(records);
+					}
+
+					@Override
+					public void handleWifiScanRecords(
+						List<KafkaRecord> records
+					) { /* ignored */ }
+
+					@Override
+					public void handleServiceEventRecords(
+						List<ServiceEvent> serviceEventRecords
+					) { /* ignored */ }
+				}
+			);
+		}
+	}
+
+	/** Process the list of received State records. */
+	private void handleKafkaStateRecords(List<KafkaRecord> records) {
+		long now = System.currentTimeMillis();
+		for (KafkaRecord record : records) {
+			// Drop old records
+			if (now - record.timestampMs > STATE_STALE_THRESHOLD_MS) {
+				logger.debug(
+					"Dropping old state record for {} at time {}",
```

Comment:
If `STATE_STALE_THRESHOLD_MS` is less than the frequency state messages are emitted, it looks like we will never issue a ping. I think, at least for now, this is fine. Just mentioning it.

Reply:
I'll move this into config